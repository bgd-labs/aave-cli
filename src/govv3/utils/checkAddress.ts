import { findObjectPaths } from 'find-object-paths';
import { Address, getAddress } from 'viem';
import * as addresses from '@bgd-labs/aave-address-book';

/**
 * Checks if address is listed on address-book
 * @param value
 * @param chainId
 * @returns string[] found paths to address-book addresses
 */
export function isKnownAddress(value: Address, chainId: number): string[] | void {
  // glob imports have no object properties
  // therefore we recreate the object via spread & remove addresses unrelated to the chain we are checking
  const transformedAddresses = Object.keys(addresses).reduce((acc, key) => {
    if ((addresses as any)[key].CHAIN_ID == chainId) acc[key] = { ...(addresses as any)[key] };
    return acc;
  }, {} as { [key: string]: any });
  // while on address book we have checksummed addresses tenderly returns non check summed
  // therefore we checksum the needle to have exact matches
  const results = findObjectPaths(transformedAddresses, { value: getAddress(value) });
  if (typeof results === 'string') return [results];
  return results;
}

export function findPayloadsController(chainId: number): Address | void {
  const key = Object.keys(addresses).find(
    (key) => addresses[key].CHAIN_ID === chainId && addresses[key].PAYLOADS_CONTROLLER
  );
  if (key) return addresses[key].PAYLOADS_CONTROLLER;
}
