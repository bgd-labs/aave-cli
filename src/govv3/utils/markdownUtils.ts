import { Hex, PublicClient } from 'viem';
import { CheckResult } from '../checks/types';

export function boolToMarkdown(value: boolean) {
  if (value) return `:white_check_mark:`;
  return `:sos:`;
}

/**
 * Turns a plaintext address into a link to etherscan page of that address
 * @param address to be linked
 * @param code whether to link to the code tab
 */
export function toAddressLink(address: Hex, md: boolean, client?: PublicClient): string {
  if (!client) return address;
  const link = `${client.chain?.blockExplorers?.default.url}/address/${address}`;
  if (md) return toMarkdownLink(link, address);
  return link;
}

/**
 * Turns a plaintext address into a link to etherscan page of that address
 * @param address to be linked
 * @param code whether to link to the code tab
 */
export function toTxLink(txn: Hex, md: boolean, client?: PublicClient): string {
  if (!client) return txn;
  const link = `${client.chain?.blockExplorers?.default.url}/tx/${txn}`;
  if (md) return toMarkdownLink(link, txn);
  return link;
}

export function toMarkdownLink(link: string, title?: any) {
  return `[${title || link}](${link})`;
}

export function renderCheckResult(check: { name: string }, result: CheckResult) {
  let response = `### Check: ${check.name} ${boolToMarkdown(!result.errors.length)}\n\n`;
  if (result.errors.length) response += `#### Errors\n\n${result.errors.join('\n')}\n\n`;
  if (result.warnings.length) response += `#### Warnings\n\n${result.warnings.join('\n')}\n\n`;
  if (result.info.length) response += `#### Info\n\n${result.info.join('\n')}\n\n`;
  return response;
}

export function renderUnixTime(time: number) {
  return new Date(time * 1000).toLocaleString('en-GB', { timeZone: 'UTC' });
}
