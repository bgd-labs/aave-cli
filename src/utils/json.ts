import JSONbig from "json-bigint";
import fs from "fs";
import path from "path";

export function readJson(filePath: string) {
  const content = fs.readFileSync(path.join(process.cwd(), filePath));
  return JSON.parse(
    JSON.stringify(JSONbig({ storeAsString: true }).parse(content))
  );
}
