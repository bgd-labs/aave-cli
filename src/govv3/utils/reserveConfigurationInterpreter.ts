import * as pools from '@bgd-labs/aave-address-book';
import {getBits} from '../../utils/storageSlots';

export function getDecodedReserveData(contractAddress: string, data?: bigint) {
  if (!data) return data;
  if (
    [
      pools.AaveV2EthereumAMM.POOL,
      pools.AaveV2Ethereum.POOL,
      pools.AaveV2Polygon.POOL,
      pools.AaveV2Avalanche.POOL,
    ]
      .map((address) => address.toLowerCase())
      .includes(contractAddress.toLowerCase())
  )
    return decodeReserveDataV2(data);
  return decodeReserveDataV3(data);
}

function decodeReserveDataV2(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const reserveFactor = getBits(data, 64n, 79n);
  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableBorrowingEnabled: !!stableBorrowingEnabled,
    reserveFactor,
  };
}

function decodeReserveDataV3(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableRateBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const paused = Number(getBits(data, 60n, 60n));
  const borrowingInIsolation = Number(getBits(data, 61n, 61n));
  const siloedBorrowingEnabled = Number(getBits(data, 62n, 62n));
  const flashloaningEnabled = Number(getBits(data, 63n, 63n));
  const reserveFactor = getBits(data, 64n, 79n);
  const borrowCap = getBits(data, 80n, 115n);
  const supplyCap = getBits(data, 116n, 151n);
  const liquidationProtocolFee = getBits(data, 152n, 167n);
  const eModeCategory = getBits(data, 168n, 175n);
  const unbackedMintCap = getBits(data, 176n, 211n);
  const debtCeiling = getBits(data, 212n, 251n);

  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableRateBorrowingEnabled,
    paused,
    borrowingInIsolation,
    reserveFactor,
    borrowCap,
    supplyCap,
    liquidationProtocolFee,
    eModeCategory,
    unbackedMintCap,
    debtCeiling,
    siloedBorrowingEnabled: !!siloedBorrowingEnabled,
    flashloaningEnabled: !!flashloaningEnabled,
  };
}
