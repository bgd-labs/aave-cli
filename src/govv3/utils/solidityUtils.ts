import { getAddress } from 'viem';
import { TenderlySimulationResponse } from '../../utils/tenderlyClient';

// --- Helper methods ---
/**
 * @notice Given a Tenderly contract object, generates a descriptive human-friendly name for that contract
 * @param contract Tenderly contract object to generate name from
 */
export function getContractName(contracts: TenderlySimulationResponse['contracts'] = [], address: string): string {
  // lower-case comparison as strict equality would fail on child contracts which are lower-cased when returned from tenderly
  const contract = contracts.find((c) => c.address.toLowerCase() === address.toLowerCase());
  if (!contract) return `unknown contract name at \`${getAddress(address)}\``;
  let contractName = contract?.contract_name;

  // If the contract is a token, include the full token name. This is useful in cases where the
  // token is a proxy, so the contract name doesn't give much useful information
  if (contract?.token_data?.name) contractName += ` (${contract?.token_data?.name})`;

  // if the contract is a proxy include it's child
  if (contract.standards?.includes('eip1967') && contract.child_contracts?.[0].address) {
    return `${contractName} at \`${getAddress(contract.address)}\` with implementation ${getContractName(
      contracts,
      contract.child_contracts?.[0].address
    )}`;
  }

  // Lastly, append the contract address and save it off
  return `${contractName} at \`${getAddress(contract.address)}\``;
}
