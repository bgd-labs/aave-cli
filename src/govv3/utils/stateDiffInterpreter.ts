import { Hex, PublicClient, getContract } from 'viem';
import { tenderlyDeepDiff } from './tenderlyDeepDiff';
import { ERC20_ABI } from '../abis/ERC20';
import * as pools from '@bgd-labs/aave-address-book';
import { getBits } from '../../utils/storageSlots';

export async function interpretStateChange(
  contractAddress: string,
  name: string = '',
  original: Record<string, any>,
  dirty: Record<string, any>,
  key: Hex,
  publicClient: PublicClient
) {
  if (name === '_reserves' && (original.configuration.data || dirty.configuration.data))
    return await reserveConfigurationChanged(contractAddress, original, dirty, key, publicClient);
  return undefined;
}

async function reserveConfigurationChanged(
  contractAddress: string,
  original: Record<string, any>,
  dirty: Record<string, any>,
  key: Hex,
  publicClient: PublicClient
) {
  const configurationBefore = getDecodedReserveData(contractAddress, original.configuration.data);
  const configurationAfter = getDecodedReserveData(contractAddress, dirty.configuration.data);
  let symbol = 'unknown';
  try {
    const erc20Contract = getContract({ publicClient, address: key, abi: ERC20_ABI });
    symbol = await erc20Contract.read.symbol();
  } catch (e) {}
  // const symbol =
  return `# decoded configuration.data for key \`${key}\` (symbol: ${symbol})
  ${tenderlyDeepDiff(configurationBefore, configurationAfter, 'configuration.data')}`;
}

function getDecodedReserveData(contractAddress: string, data?: bigint) {
  if (!data) return {};
  if (
    [pools.AaveV2EthereumAMM.POOL, pools.AaveV2Ethereum.POOL, pools.AaveV2Polygon.POOL, pools.AaveV2Avalanche.POOL]
      .map((address) => address.toLowerCase())
      .includes(contractAddress.toLowerCase())
  )
    return decodeReserveDataV2(data);
  return decodeReserveDataV3(data);
}

function decodeReserveDataV2(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const reserveFactor = getBits(data, 64n, 79n);
  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    reserveFactor,
  };
}

function decodeReserveDataV3(data: bigint) {
  const ltv = getBits(data, 0n, 15n);
  const liquidationThreshold = getBits(data, 16n, 31n);
  const liquidationBonus = getBits(data, 32n, 47n);
  const decimals = getBits(data, 48n, 55n);
  const active = Number(getBits(data, 56n, 56n));
  const frozen = Number(getBits(data, 57n, 57n));
  const borrowingEnabled = Number(getBits(data, 58n, 58n));
  const stableRateBorrowingEnabled = Number(getBits(data, 59n, 59n));
  const paused = Number(getBits(data, 60n, 60n));
  const borrowingInIsolation = Number(getBits(data, 61n, 61n));
  const siloedBorrowingEnabled = Number(getBits(data, 62n, 62n));
  const flashloaningEnabled = Number(getBits(data, 63n, 63n));
  const reserveFactor = getBits(data, 64n, 79n);
  const borrowCap = getBits(data, 80n, 115n);
  const supplyCap = getBits(data, 116n, 151n);
  const liquidationProtocolFee = getBits(data, 152n, 167n);
  const eModeCategory = getBits(data, 168n, 175n);
  const unbackedMintCap = getBits(data, 176n, 211n);
  const debtCeiling = getBits(data, 212n, 251n);

  return {
    ltv,
    liquidationThreshold,
    liquidationBonus,
    decimals,
    active: !!active,
    frozen: !!frozen,
    borrowingEnabled: !!borrowingEnabled,
    stableRateBorrowingEnabled,
    paused,
    borrowingInIsolation,
    reserveFactor,
    borrowCap,
    supplyCap,
    liquidationProtocolFee,
    eModeCategory,
    unbackedMintCap,
    debtCeiling,
    siloedBorrowingEnabled: !!siloedBorrowingEnabled,
    flashloaningEnabled: !!flashloaningEnabled,
  };
}
