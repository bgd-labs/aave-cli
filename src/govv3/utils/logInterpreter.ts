import {type Client, type Hex, formatUnits} from 'viem';
import type {Input} from '../../utils/tenderlyClient';
import {findAsset} from './checkAddress';
import {formatNumberString, prettifyNumber} from './markdownUtils';

// events emitted typically on the erc20
const tokenAmountEvents = ['Transfer', 'Approval', 'Burn', 'Mint', 'BalanceTransfer'];

// events emitted on the pool
const reserveEvents = ['Withdraw', 'Supply', 'Deposit'];

export async function interpretLog(
  client: Client,
  address: Hex,
  name: string | null,
  inputs: Input[],
) {
  if (name && tokenAmountEvents.includes(name)) {
    // fields formatted by the asset decimal
    const decimalFieldNames = ['value', 'amount', 'wad'];
    for (const name of decimalFieldNames) {
      const valueIndex = inputs.findIndex((i) => i.soltype!.name === name);
      if (valueIndex !== -1) {
        const asset = await findAsset(client, address);
        if (asset) {
          inputs[valueIndex].value = prettifyNumber({
            value: inputs[valueIndex].value as string,
            decimals: asset.decimals,
          });
        }
      }
    }
  }
  if (name && reserveEvents.includes(name)) {
    const valueIndex = inputs.findIndex((i) => i.soltype!.name === 'amount');
    const reserveIndex = inputs.findIndex((i) => i.soltype!.name === 'reserve');
    if (valueIndex !== -1 && reserveIndex !== -1) {
      const asset = await findAsset(client, inputs[reserveIndex].value as Hex);
      if (asset) {
        inputs[valueIndex].value = prettifyNumber({
          value: inputs[valueIndex].value as string,
          decimals: asset.decimals,
        });
      }
    }
  }
  if (name && ['Mint', 'ReserveDataUpdated', 'Burn'].includes(name)) {
    // fields formatted by the asset decimal
    const decimalFieldNames = ['liquidityIndex', 'variableBorrowIndex', 'index'];
    for (const name of decimalFieldNames) {
      const valueIndex = inputs.findIndex((i) => i.soltype!.name === name);
      if (valueIndex !== -1) {
        inputs[valueIndex].value = prettifyNumber({
          value: inputs[valueIndex].value as string,
          decimals: 27,
        });
      }
    }
  }
  const parsedInputs = inputs
    ?.map(
      (i) =>
        `${i.soltype!.name}: ${typeof i.value === 'object' ? JSON.stringify(i.value) : i.value}`,
    )
    .join(', ');
  return `  - \`${name}(${parsedInputs || ''})\``;
}
