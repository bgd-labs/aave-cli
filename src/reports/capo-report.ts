import {formatUnits} from 'viem';
import {formatTimestamp} from '../govv3/utils/markdownUtils';
import {CapoSnapshot} from './snapshot-types';

type Price = {
  sourcePrice: string;
  referencePrice: string;
  diff: string;
  dayToDayGrowth: string;
  smoothedGrowth: string;
  date: string;
};

export async function generateCapoReport(snapshot: CapoSnapshot) {
  // map to dates and formatted values

  let prices: Price[] = [];

  let maxDayToDayGrowth = 0;
  let maxSmoothedGrowth = 0;

  for (const key in snapshot.prices) {
    const price = snapshot.prices[key];

    const sourcePrice = formatUnits(BigInt(price.sourcePrice), snapshot.decimals);
    const referencePrice = formatUnits(BigInt(price.referencePrice), snapshot.decimals);
    const dayToDayGrowth = (price.dayToDayGrowth / 100).toFixed(2);
    const smoothedGrowth = (price.smoothedGrowth / 100).toFixed(2);
    const diff = (
      (100 * (Number(sourcePrice) - Number(referencePrice))) /
      ((Number(sourcePrice) + Number(referencePrice)) / 2)
    ).toFixed(2);

    const formattedDate = formatTimestamp(price.timestamp);

    prices.push({
      sourcePrice,
      referencePrice,
      diff,
      dayToDayGrowth,
      smoothedGrowth,
      date: formattedDate,
    });

    maxDayToDayGrowth = Math.max(maxDayToDayGrowth, price.dayToDayGrowth);
    maxSmoothedGrowth = Math.max(maxSmoothedGrowth, price.smoothedGrowth);
  }

  prices = prices.slice(snapshot.minSnapshotDelay);

  // generate md report
  let content = '';
  content += `# Capo Report\n\n`;
  content += `| ${snapshot.source} | ${snapshot.reference} | Diff | Date | ${snapshot.minSnapshotDelay}-day growth in yearly % |\n`;
  content += `| --- | --- | --- | --- | --- |\n`;
  prices.forEach((price) => {
    content += `| ${price.sourcePrice} | ${price.referencePrice} | ${price.diff}% | ${price.date} | ${price.smoothedGrowth}% |\n`;
  });
  content += `\n\n`;
  content += `* ${snapshot.minSnapshotDelay}-day growth is calculated as an annualized percentage relative to the value of the rate ${snapshot.minSnapshotDelay} days prior. \n`;
  const maxYearlyGrowthPercent = (snapshot.maxYearlyGrowthPercent / 100).toFixed(2);
  const maxDayToDayGrowthPercent = (maxDayToDayGrowth / 100).toFixed(2);
  const maxSmoothedGrowthPercent = (maxSmoothedGrowth / 100).toFixed(2);

  content += `\n\n`;
  content += `| Max Yearly % | Max Day-to-day yearly % | Max ${snapshot.minSnapshotDelay}-day yearly % | \n`;
  content += `| --- | --- | --- |\n`;
  content += `| ${maxYearlyGrowthPercent}% | ${maxDayToDayGrowthPercent}% | ${maxSmoothedGrowthPercent}% | \n`;
  content += `\n\n`;

  content += `* Max day-to-day yearly % indicates the maximum growth between two emissions as an annualized percentage. \n`;

  return content;
}
