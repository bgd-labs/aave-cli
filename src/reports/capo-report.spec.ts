import {describe, expect, it} from 'vitest';
import {readJsonFile} from '../utils/json';
import {generateCapoReport} from './capo-report';

describe('capo report', () => {
  it(
    'should generate a well formatted capo report',
    async () => {
      const snapshot = readJsonFile('/src/reports/mocks/capo.json');
      const content = await generateCapoReport(snapshot);
      expect(content).toMatchSnapshot();
    },
    {timeout: 30000},
  );
});
