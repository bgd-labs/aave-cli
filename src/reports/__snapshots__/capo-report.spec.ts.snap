// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`capo report > should generate a well formatted capo report 1`] = `
"# Capo Report

| Capped weETH / eETH(ETH) / USD | ETH / USD | Diff | Date | 7-day growth in yearly % |
| --- | --- | --- | --- | --- |
| 4022.69511587 | 3892.37199 | 3.29% | 10 Mar 2024 | 3.39% |
| 4187.79647706 | 4051.72157013 | 3.30% | 11 Mar 2024 | 3.65% |
| 4084.9163739 | 3951.8 | 3.31% | 12 Mar 2024 | 3.45% |
| 4142.24713661 | 4006.96759142 | 3.32% | 13 Mar 2024 | 3.51% |
| 4013.18898648 | 3881.86606 | 3.33% | 15 Mar 2024 | 3.19% |
| 3890.97413303 | 3763.4063 | 3.33% | 16 Mar 2024 | 2.99% |
| 3657.27699156 | 3537.15488 | 3.34% | 17 Mar 2024 | 2.78% |
| 3730.51277856 | 3607.7 | 3.35% | 18 Mar 2024 | 2.75% |
| 3527.1199042 | 3410.78 | 3.35% | 19 Mar 2024 | 2.58% |
| 3358.22435506 | 3247.26491 | 3.36% | 20 Mar 2024 | 2.38% |
| 3639.21625509 | 3518.59 | 3.37% | 21 Mar 2024 | 2.56% |
| 3626.78329924 | 3506.32167007 | 3.38% | 22 Mar 2024 | 2.58% |
| 3443.40356942 | 3328.75483 | 3.39% | 23 Mar 2024 | 2.68% |
| 3432.83895716 | 3318.29815289 | 3.39% | 24 Mar 2024 | 2.74% |
| 3589.63867272 | 3469.62555434 | 3.40% | 25 Mar 2024 | 2.69% |
| 3785.95083613 | 3659.13658361 | 3.41% | 26 Mar 2024 | 2.69% |
| 3690.05083183 | 3566.1009 | 3.42% | 27 Mar 2024 | 2.88% |
| 3699.94503716 | 3575.39 | 3.42% | 28 Mar 2024 | 2.71% |
| 3658.21925515 | 3534.79 | 3.43% | 29 Mar 2024 | 2.74% |
| 3635.19779513 | 3512.3 | 3.44% | 30 Mar 2024 | 2.67% |
| 3732.54580315 | 3606.17191378 | 3.44% | 31 Mar 2024 | 2.56% |
| 3663.60436472 | 3539.35545849 | 3.45% | 01 Apr 2024 | 2.51% |
| 3438.73281666 | 3321.7664498 | 3.46% | 02 Apr 2024 | 2.71% |
| 3426.86187101 | 3310.09407509 | 3.47% | 03 Apr 2024 | 2.53% |


* 7-day growth is calculated as an annualized percentage relative to the value of the rate 7 days prior. 


| Max Yearly % | Max Day-to-day yearly % | Max 7-day yearly % | 
| --- | --- | --- | --- |
| 8.75% | 4.91% | 3.65% | 


* Max day-to-day yearly % indicates the maximum growth between two emissions as an annualized percentage. 
"
`;
